src/Interpreters/ArithmeticOperationsInAgrFuncOptimize.cpp

```diff
@@ -0,0 +1,286 @@
+        return fl;
+    }
+    else if ((func_node->arguments->children[0]->as<ASTFunction>() && inter_func_name != func_node->arguments->children[0]->as<ASTFunction>()->name) ||
+            (func_node->arguments->children.size() == 2 && func_node->arguments->children[1]->as<ASTFunction>() &&
+            inter_func_name != func_node->arguments->children[1]->as<ASTFunction>()->name))
+    {
+        if (func_node->arguments->children[0]->as<ASTFunction>())
+            return {{func_node->arguments->children[1]}, {func_node->arguments->children[0]}};
+        else
+            return {{func_node->arguments->children[0]}, {func_node->arguments->children[1]}};
+    }
+    else
+    {
+        std::pair<ASTs, ASTs> fl = tryGetConst(func_node->name, func_node->arguments->children);
+        if (!fl.second[0]->as<ASTFunction>())
+        {
+            std::pair<ASTs, ASTs> ans = {fl.first, fl.second};
+            return ans;
+        }
+
+        std::pair<ASTs, ASTs> ans = findAllConsts(fl.second[0]->as<ASTFunction>(), inter_func_name);
+
+        if (fl.first.size() == 1)
+        {
+            if (!fl.second[0]->as<ASTFunction>())
+                ans.second.push_back(fl.second[0]);
+
+            ans.first.push_back(fl.first[0]);
+        }
+        else if (fl.first.empty())
+        {
+            if (!fl.second[0]->as<ASTFunction>() || fl.second[0]->as<ASTFunction>()->name != inter_func_name)
+                ans.second.push_back(fl.second[0]);
+
+            if (fl.second.size() == 2 && (!fl.second[1]->as<ASTFunction>() || fl.second[1]->as<ASTFunction>()->name != inter_func_name))
+                ans.second.push_back(fl.second[1]);
+        }
+        else
+            throw Exception("did not expect that", ErrorCodes::UNEXPECTED_AST_STRUCTURE);
+        return ans;
+    }
+}
+
+/// rebuilds tree, all scalar values now outside the main func
+void buildTree(ASTFunction * old_tree, std::string& func_name, std::string& intro_func, std::pair<ASTs, ASTs> & tree_comp, std::pair<int, std::string> & flag)
+{
+    ASTFunction copy = *old_tree;
+    ASTs cons_val = tree_comp.first;
+    ASTs non_cons = tree_comp.second;
+
+    old_tree->name = intro_func;
+    for (auto & i : cons_val)
+    {
+        old_tree->arguments->children = {};
+        old_tree->arguments->children.push_back(i);
+
+        old_tree->arguments->children.push_back(makeASTFunction(intro_func));
+        old_tree = old_tree->arguments->children[1]->as<ASTFunction>();
+    }
+
+    if (flag.first == -1)
+        old_tree->name = flag.second;
+    else
+        old_tree->name = func_name;
+
+    if (non_cons.empty())
+        throw Exception("Aggregate function " + func_name + " requires single argument",
+                        ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+    if (non_cons.size() == 1)
+        old_tree->arguments->children.push_back(non_cons[0]);
+    else
+    {
+        size_t i = 0;
+
+        while (i < non_cons.size() - 2)
```

Comment:
This cycle is some dangerous opeation with pointers. It should be moved outside in separate function to understand it reason (need good function name) and side effects. For now it's not clear why we do it and why we could do it without care.